power_dashboard/eia_api.py

```diff
@@ -0,0 +1,300 @@
+    generation_types_by_ba = get_eia_grid_mix_timeseries_hourly(all_source_bas
+        , start_date=start_date
+        , end_date=end_date
+        , frequency="hourly"
+        ).rename(
+        {"respondent": "fromba", "type-name": "generation_type"}, axis="columns"
+    )
+    # The goal is to get a DataFrame of energy used at the local BA (in MWh), broken down by both
+    #  * the BA that the energy came from, and 
+    #  * the fuel type of that energy.
+    # So we'll end up with one row for each combination of source BA and fuel type.
+
+    # To get there, we need to combine the amount of imported energy from each source ba with grid mix for that source BA.
+    # The general formula is:
+    # Power consumed locally from a (BA, fuel type) combination = 
+    #    total power consumed locally from this source BA * (fuel type as a % of source BA's generation)
+    # fuel type as a % of source BA's generation = 
+    #    (total generation at source BA) / (total generation for this fuel type at this BA)
+    total_generation_by_source_ba = generation_types_by_ba.groupby(["period", "fromba"])[
+        "Generation (MWh)"
+    ].sum()
+    generation_types_by_ba_with_totals = generation_types_by_ba.join(
+        total_generation_by_source_ba,
+        how="left",
+        on=["period", "fromba"],
+        rsuffix=" Total",
+    )
+    generation_types_by_ba_with_totals["Generation (% of BA generation)"] = (
+        generation_types_by_ba_with_totals["Generation (MWh)"]
+        / generation_types_by_ba_with_totals["Generation (MWh) Total"]
+    )
+    generation_types_by_ba_with_totals_and_source_ba_breakdown = generation_types_by_ba_with_totals.merge(
+        energy_consumed_locally_by_source_ba.rename(
+            {"Power consumed locally (MWh)": "Power consumed locally from source BA (MWh)"},
+            axis="columns",
+        ),
+        on=["period", "fromba"],
+    )
+    generation_types_by_ba_with_totals_and_source_ba_breakdown = generation_types_by_ba_with_totals_and_source_ba_breakdown.loc[generation_types_by_ba_with_totals_and_source_ba_breakdown['Generation (MWh) Total'] > 0]
+    #full_df_reindexed = (
+    #    generation_types_by_ba_with_totals_and_source_ba_breakdown.set_index(
+    #        ["period", "fromba", "generation_type"]
+    #    )
+    #)
+    co2_kwh_est = generation_types_by_ba_with_totals_and_source_ba_breakdown
+    co2_kwh_est['CO2/(kWh)'] = co2_kwh_est[['fueltype', 'Generation (% of BA generation)', "Power consumed locally from source BA (MWh)", "Generation (MWh) Total"]].apply(co2_contrib, axis=1);
+    co2_kwh_est_sum = (
+        co2_kwh_est.groupby(["timestamp"])[
+            "CO2/(kWh)"
+        ].sum()
+    ).reset_index()
+    #st.text(co2_kwh_est_sum.dtypes)
+    co2_kwh_est_sum["timestamp"] = co2_kwh_est_sum["timestamp"].apply(
+        pd.to_datetime, format="%Y/%m/%d %H:%M:%S", utc=True
+    )
+    #st.text(co2_kwh_est_sum.dtypes)
+
+    return co2_kwh_est_sum
```

Comment:
The gridemissions API is only serving data before 6/15 in the bulk files and from 7/20 and 7/21 via API. However, I was able to run his pipeline to generate bulk files covering 6/15 to present. Ideally, I would load this to the gridemissions-ts table and we could pull it from there, so the same co2 intensity algorithm is used consistently throughout. Not a blocker for wrapping up our submission, though!